# 시각화!

더 뽑아먹을 거리가 있는지 찾아보기 위해 오늘도 데이터를 살펴봅니다.


In [19]:
import ast
import itertools
from operator import itemgetter
from pprint import pprint

import numpy as np
import pandas as pd

In [2]:
chars = pd.read_excel('./datasets/characters.xlsx')
chars.info()
chars.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 452 entries, 0 to 451
Data columns (total 24 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   451 non-null    object 
 1   name_ko              452 non-null    object 
 2   name_en              452 non-null    object 
 3   special              452 non-null    bool   
 4   gender               452 non-null    object 
 5   gender_asia          452 non-null    object 
 6   species              391 non-null    object 
 7   birthday             452 non-null    object 
 8   birthday_month       452 non-null    int64  
 9   birthday_day         452 non-null    int64  
 10  birthday_yday        452 non-null    int64  
 11  hobby                391 non-null    object 
 12  personality          391 non-null    object 
 13  personality_subtype  391 non-null    object 
 14  colors               391 non-null    object 
 15  styles               391 non-null    obj

id name_ko  name_en  special  gender gender_asia species birthday  \
0  admiral      일섭  Admiral    False    Male        Male       새    01-27   
1   agents      2호  Agent S    False  Female      Female     다람쥐    07-02   
2    agnes    아그네스    Agnes    False  Female      Female      돼지    04-21   
3       al      우락       Al    False    Male        Male     고릴라    10-18   
4  alfonso    알베르트  Alfonso    False    Male        Male      악어    06-09   

   birthday_month  birthday_day  ...          colors          styles tier  \
0               1            27  ...  ['검정색', '파랑색']           ['쿨']  6.0   
1               7             2  ...  ['검정색', '파랑색']   ['심플', '액티브']  6.0   
2               4            21  ...  ['하양색', '핑크색']  ['심플', '엘레강스']  4.0   
3              10            18  ...  ['하양색', '빨강색']         ['액티브']  6.0   
4               6             9  ...  ['빨강색', '파랑색']          ['심플']  6.0   

   rank catchphrase                                favorite_saying  \
0  89.0     aye aye                         Only quitters give up.   
1  45.0    sidekick          You gotta put the pedal to the metal!   
2  26.0     snuffle                         You reap what you sow.   
3  19.0     ayyyeee  Life is a buffet. Always go back for seconds.   
4   6.0    it'sa me                 Slow and steady wins the race.   

   favorite_song                                        image_house  \
0     Steep Hill  https://acnhcdn.com/drivesync/render/houses/br...   
1  Go K.K. Rider  https://acnhcdn.com/drivesync/render/houses/sq...   
2     K.K. House  https://acnhcdn.com/drivesync/render/houses/pi...   
3  Go K.K. Rider  https://acnhcdn.com/drivesync/render/houses/go...   
4    Forest Life  https://acnhcdn.com/drivesync/render/houses/cr...   

                                     image_icon  \
0  https://acnhcdn.com/latest/NpcIcon/brd06.png   
1  https://acnhcdn.com/latest/NpcIcon/squ05.png   
2  https://acnhcdn.com/latest/NpcIcon/pig17.png   
3  https://acnhcdn.com/latest/NpcIcon/gor08.png   
4  https://acnhcdn.com/latest/NpcIcon/crd00.png   

                                         image_photo  
0  https://acnhcdn.com/latest/NpcBromide/NpcNmlBr...  
1  https://acnhcdn.com/latest/NpcBromide/NpcNmlSq...  
2  https://acnhcdn.com/latest/NpcBromide/NpcNmlPi...  
3  https://acnhcdn.com/latest/NpcBromide/NpcNmlGo...  
4  https://acnhcdn.com/latest/NpcBromide/NpcNmlCr...  

[5 rows x 24 columns]

## 뭘 할 수 있는지 보기

생각하기 귀찮으니까 컴퓨터한테 시킵니다.


In [46]:
# list(itertools.permutations(['성격', '취미', '색깔', '성별', '스타일'], 2))
list(itertools.permutations(['성격', '취미', '성별',], 2))

[('성격', '취미'),
 ('성격', '성별'),
 ('취미', '성격'),
 ('취미', '성별'),
 ('성별', '성격'),
 ('성별', '취미')]

## 성격별 xx

성격별로 뭔가 달라지는 게 있을까요


### 성격별 취미

```
personality  hobby
느끼함          교육       11
             운동        4
             음악       14
             자연        5
단순 활발        놀이        7
             운동        6
             음악       11
먹보           교육        1
             놀이       32
             운동        4
             자연       23
무뚝뚝          교육       19
             놀이        2
             운동       10
             음악       13
             자연       11
성숙함          교육       10
             운동        3
             음악        8
             자연        2
             패션       32
아이돌          놀이        9
             운동        3
             음악        5
             자연        1
             패션       31
운동광          교육        1
             놀이       15
             운동       36
             음악        3
친절함          교육       22
             음악       10
             자연       24
             패션        3
dtype: int64
```

In [10]:
chars.groupby(['personality', 'hobby']).size()

personality  hobby
느끼함          교육       11
             운동        4
             음악       14
             자연        5
단순 활발        놀이        7
             운동        6
             음악       11
먹보           교육        1
             놀이       32
             운동        4
             자연       23
무뚝뚝          교육       19
             놀이        2
             운동       10
             음악       13
             자연       11
성숙함          교육       10
             운동        3
             음악        8
             자연        2
             패션       32
아이돌          놀이        9
             운동        3
             음악        5
             자연        1
             패션       31
운동광          교육        1
             놀이       15
             운동       36
             음악        3
친절함          교육       22
             음악       10
             자연       24
             패션        3
dtype: int64

### 성격별 성별

```
personality  gender_asia
느끼함          Male           34
단순 활발        Female         24
먹보           Male           60
무뚝뚝          Male           55
성숙함          Female         55
아이돌          Female         49
운동광          Male           55
친절함          Female         59
dtype: int64
```

In [50]:
chars.groupby(['personality', 'gender_asia']).size()

personality  gender_asia
느끼함          Male           34
단순 활발        Female         24
먹보           Male           60
무뚝뚝          Male           55
성숙함          Female         55
아이돌          Female         49
운동광          Male           55
친절함          Female         59
dtype: int64

### 취미별 성격

```
hobby  personality
교육     느끼함            11
       먹보              1
       무뚝뚝            19
       성숙함            10
       운동광             1
       친절함            22
놀이     단순 활발           7
       먹보             32
       무뚝뚝             2
       아이돌             9
       운동광            15
운동     느끼함             4
       단순 활발           6
       먹보              4
       무뚝뚝            10
       성숙함             3
       아이돌             3
       운동광            36
음악     느끼함            14
       단순 활발          11
       무뚝뚝            13
       성숙함             8
       아이돌             5
       운동광             3
       친절함            10
자연     느끼함             5
       먹보             23
       무뚝뚝            11
       성숙함             2
       아이돌             1
       친절함            24
패션     성숙함            32
       아이돌            31
       친절함             3
dtype: int64
```


In [52]:
chars.groupby(['hobby', 'personality']).size()


hobby  personality
교육     느끼함            11
       먹보              1
       무뚝뚝            19
       성숙함            10
       운동광             1
       친절함            22
놀이     단순 활발           7
       먹보             32
       무뚝뚝             2
       아이돌             9
       운동광            15
운동     느끼함             4
       단순 활발           6
       먹보              4
       무뚝뚝            10
       성숙함             3
       아이돌             3
       운동광            36
음악     느끼함            14
       단순 활발          11
       무뚝뚝            13
       성숙함             8
       아이돌             5
       운동광             3
       친절함            10
자연     느끼함             5
       먹보             23
       무뚝뚝            11
       성숙함             2
       아이돌             1
       친절함            24
패션     성숙함            32
       아이돌            31
       친절함             3
dtype: int64

### 취미별 성별

```
hobby  gender_asia
교육     Female         32
       Male           32
놀이     Female         16
       Male           49
운동     Female         12
       Male           54
음악     Female         34
       Male           30
자연     Female         27
       Male           39
패션     Female         66
dtype: int64
```

In [56]:
chars.groupby(['hobby', 'gender_asia']).size()

hobby  gender_asia
교육     Female         32
       Male           32
놀이     Female         16
       Male           49
운동     Female         12
       Male           54
음악     Female         34
       Male           30
자연     Female         27
       Male           39
패션     Female         66
dtype: int64

### 성별별 취미

```
gender_asia  hobby
Female       교육       32
             놀이       16
             운동       12
             음악       34
             자연       27
             패션       66
Male         교육       32
             놀이       49
             운동       54
             음악       30
             자연       39
dtype: int64
```

In [53]:
chars.groupby(['gender_asia', 'hobby']).size()

gender_asia  hobby
Female       교육       32
             놀이       16
             운동       12
             음악       34
             자연       27
             패션       66
Male         교육       32
             놀이       49
             운동       54
             음악       30
             자연       39
dtype: int64

### 성별별 성격

이게 의미가 있을 수도 있는 이유는 성별별로 성격이 따로있다는 거는 게임을 안해보면 절대 상상도 못할 일이기 때문입니다.

```
gender_asia  personality
Female       단순 활발          24
             성숙함            55
             아이돌            49
             친절함            59
Male         느끼함            34
             먹보             60
             무뚝뚝            55
             운동광            55
dtype: int64
```

In [54]:
chars.groupby(['gender_asia', 'personality']).size()

gender_asia  personality
Female       단순 활발          24
             성숙함            55
             아이돌            49
             친절함            59
Male         느끼함            34
             먹보             60
             무뚝뚝            55
             운동광            55
dtype: int64

### 성별별 취미

```
gender_asia  hobby
Female       교육       32
             놀이       16
             운동       12
             음악       34
             자연       27
             패션       66
Male         교육       32
             놀이       49
             운동       54
             음악       30
             자연       39
dtype: int64
```

In [55]:
chars.groupby(['gender_asia', 'hobby']).size()

gender_asia  hobby
Female       교육       32
             놀이       16
             운동       12
             음악       34
             자연       27
             패션       66
Male         교육       32
             놀이       49
             운동       54
             음악       30
             자연       39
dtype: int64

## 인기도?

tier와 rank 2가지가 있는데 rank까지는 쓸 필요가 없었으면 합니다.
너무 머리 아프니까...


### 성별 분포

깜박할뻔 했는데 성별의 분포는 해석에 꽤 중요합니다.


In [57]:
chars.groupby('gender_asia').size()

gender_asia
Female    205
Male      247
dtype: int64

남 247 : 여 205

근데 지금 스페셜 주민도 끼어 있습니다.

얘네는 생일 말곤 속성이 없어서 도움이 안됩니다.


In [63]:
chars_v = pd.read_excel('./datasets/characters-similarity.xlsx')
chars_v['colors'] = chars_v['colors'].apply(ast.literal_eval)
chars_v['styles'] = chars_v['styles'].apply(ast.literal_eval)
chars_v.head()

id  name_en name_ko                                    image_icon  \
0  admiral  Admiral      일섭  https://acnhcdn.com/latest/NpcIcon/brd06.png   
1   agents  Agent S      2호  https://acnhcdn.com/latest/NpcIcon/squ05.png   
2    agnes    Agnes    아그네스  https://acnhcdn.com/latest/NpcIcon/pig17.png   
3       al       Al      우락  https://acnhcdn.com/latest/NpcIcon/gor08.png   
4  alfonso  Alfonso    알베르트  https://acnhcdn.com/latest/NpcIcon/crd00.png   

     species  gender birthday  birthday_month  birthday_day personality hobby  \
0       Bird    Male    01-27               1            27         무뚝뚝    자연   
1   Squirrel  Female    07-02               7             2         아이돌    운동   
2        Pig  Female    04-21               4            21       단순 활발    놀이   
3    Gorilla    Male    10-18              10            18          먹보    운동   
4  Alligator    Male    06-09               6             9          먹보    놀이   

       styles      colors  
0         [쿨]  [검정색, 파랑색]  
1   [심플, 액티브]  [검정색, 파랑색]  
2  [심플, 엘레강스]  [하양색, 핑크색]  
3       [액티브]  [하양색, 빨강색]  
4        [심플]  [빨강색, 파랑색]

In [69]:
chars_v.groupby('gender').size()

gender
Female    187
Male      204
dtype: int64

f: 187 vs. m: 204

대체로 비슷비슷한 것 같습니다.


### 성별/인기도

여자가 남자보다 약간 적은 것을 감안하면 여자일수록 약간 인기가 많은 경향이 있습니다.


In [68]:
chars_v = chars[chars['special'].eq(False)]

In [70]:
chars_v.groupby(['gender', 'tier']).size()

gender  tier
Female  1.0      6
        2.0     14
        3.0     18
        4.0     30
        5.0     58
        6.0     61
Male    1.0      6
        2.0     10
        3.0     11
        4.0     25
        5.0     55
        6.0     96
dtype: int64

In [71]:
chars_v.groupby(['tier', 'gender']).size()

tier  gender
1.0   Female     6
      Male       6
2.0   Female    14
      Male      10
3.0   Female    18
      Male      11
4.0   Female    30
      Male      25
5.0   Female    58
      Male      55
6.0   Female    61
      Male      96
dtype: int64

### 성격 / 인기도

성격과 인기도는 뚜렷한 상관 관계는 없는 것 같습니다.


```
personality  tier
느끼함          1.0      2
             2.0      2
             3.0      2
             4.0      6
             5.0      5
             6.0     17
단순 활발        2.0      1
             3.0      5
             4.0      6
             5.0      8
             6.0      4
먹보           1.0      4
             2.0      4
             3.0      2
             4.0      9
             5.0     19
             6.0     21
무뚝뚝          2.0      2
             3.0      4
             4.0      5
             5.0     13
             6.0     31
성숙함          1.0      2
             2.0      1
             3.0      1
             4.0      6
             5.0     15
             6.0     30
아이돌          2.0      6
             3.0      4
             4.0     10
             5.0     16
             6.0     13
운동광          2.0      2
             3.0      3
             4.0      5
             5.0     18
             6.0     27
친절함          1.0      4
             2.0      6
             3.0      8
             4.0      8
             5.0     19
             6.0     14
dtype: int64
```

```
tier  personality
1.0   느끼함             2
      먹보              4
      성숙함             2
      친절함             4
2.0   느끼함             2
      단순 활발           1
      먹보              4
      무뚝뚝             2
      성숙함             1
      아이돌             6
      운동광             2
      친절함             6
3.0   느끼함             2
      단순 활발           5
      먹보              2
      무뚝뚝             4
      성숙함             1
      아이돌             4
      운동광             3
      친절함             8
4.0   느끼함             6
      단순 활발           6
      먹보              9
      무뚝뚝             5
      성숙함             6
      아이돌            10
      운동광             5
      친절함             8
5.0   느끼함             5
      단순 활발           8
      먹보             19
      무뚝뚝            13
      성숙함            15
      아이돌            16
      운동광            18
      친절함            19
6.0   느끼함            17
      단순 활발           4
      먹보             21
      무뚝뚝            31
      성숙함            30
      아이돌            13
      운동광            27
      친절함            14
dtype: int64
```

In [72]:
chars_v.groupby(['personality', 'tier']).size()

personality  tier
느끼함          1.0      2
             2.0      2
             3.0      2
             4.0      6
             5.0      5
             6.0     17
단순 활발        2.0      1
             3.0      5
             4.0      6
             5.0      8
             6.0      4
먹보           1.0      4
             2.0      4
             3.0      2
             4.0      9
             5.0     19
             6.0     21
무뚝뚝          2.0      2
             3.0      4
             4.0      5
             5.0     13
             6.0     31
성숙함          1.0      2
             2.0      1
             3.0      1
             4.0      6
             5.0     15
             6.0     30
아이돌          2.0      6
             3.0      4
             4.0     10
             5.0     16
             6.0     13
운동광          2.0      2
             3.0      3
             4.0      5
             5.0     18
             6.0     27
친절함          1.0      4
             2.0      6
             3.0      

In [73]:
chars_v.groupby(['tier', 'personality']).size()

tier  personality
1.0   느끼함             2
      먹보              4
      성숙함             2
      친절함             4
2.0   느끼함             2
      단순 활발           1
      먹보              4
      무뚝뚝             2
      성숙함             1
      아이돌             6
      운동광             2
      친절함             6
3.0   느끼함             2
      단순 활발           5
      먹보              2
      무뚝뚝             4
      성숙함             1
      아이돌             4
      운동광             3
      친절함             8
4.0   느끼함             6
      단순 활발           6
      먹보              9
      무뚝뚝             5
      성숙함             6
      아이돌            10
      운동광             5
      친절함             8
5.0   느끼함             5
      단순 활발           8
      먹보             19
      무뚝뚝            13
      성숙함            15
      아이돌            16
      운동광            18
      친절함            19
6.0   느끼함            17
      단순 활발           4
      먹보             21
      무뚝뚝            31
      성숙함            3

### 취미 / 인기도

취미별 인기도도 뚜렷한 경향은 없습니다.
얼핏 보면 자연이 꽤 많아 보이기도 하지만 확신할 수준은 아닙니다.

```
hobby  tier
교육     1.0      2
       2.0      5
       3.0      4
       4.0     10
       5.0     13
       6.0     30
놀이     1.0      1
       2.0      6
       3.0      6
       4.0     12
       5.0     15
       6.0     24
운동     2.0      2
       3.0      2
       4.0      4
       5.0     20
       6.0     38
음악     1.0      3
       2.0      7
       3.0      6
       4.0      9
       5.0     19
       6.0     20
자연     1.0      6
       2.0      3
       3.0      8
       4.0     10
       5.0     19
       6.0     20
패션     2.0      1
       3.0      3
       4.0     10
       5.0     27
       6.0     25
dtype: int64
```

```
tier  hobby
1.0   교육        2
      놀이        1
      음악        3
      자연        6
2.0   교육        5
      놀이        6
      운동        2
      음악        7
      자연        3
      패션        1
3.0   교육        4
      놀이        6
      운동        2
      음악        6
      자연        8
      패션        3
4.0   교육       10
      놀이       12
      운동        4
      음악        9
      자연       10
      패션       10
5.0   교육       13
      놀이       15
      운동       20
      음악       19
      자연       19
      패션       27
6.0   교육       30
      놀이       24
      운동       38
      음악       20
      자연       20
      패션       25
dtype: int64
```

In [74]:
chars_v.groupby(['hobby', 'tier']).size()

hobby  tier
교육     1.0      2
       2.0      5
       3.0      4
       4.0     10
       5.0     13
       6.0     30
놀이     1.0      1
       2.0      6
       3.0      6
       4.0     12
       5.0     15
       6.0     24
운동     2.0      2
       3.0      2
       4.0      4
       5.0     20
       6.0     38
음악     1.0      3
       2.0      7
       3.0      6
       4.0      9
       5.0     19
       6.0     20
자연     1.0      6
       2.0      3
       3.0      8
       4.0     10
       5.0     19
       6.0     20
패션     2.0      1
       3.0      3
       4.0     10
       5.0     27
       6.0     25
dtype: int64

In [75]:
chars_v.groupby(['tier', 'hobby']).size()

tier  hobby
1.0   교육        2
      놀이        1
      음악        3
      자연        6
2.0   교육        5
      놀이        6
      운동        2
      음악        7
      자연        3
      패션        1
3.0   교육        4
      놀이        6
      운동        2
      음악        6
      자연        8
      패션        3
4.0   교육       10
      놀이       12
      운동        4
      음악        9
      자연       10
      패션       10
5.0   교육       13
      놀이       15
      운동       20
      음악       19
      자연       19
      패션       27
6.0   교육       30
      놀이       24
      운동       38
      음악       20
      자연       20
      패션       25
dtype: int64

### 특수 주민은??

특수 주민은 다른 정보는 없어도 최소한 성별은 볼 수 있습니다.
성별을 검사해 봅시다.


In [76]:
chars_s = chars[chars['special'].eq(True)]

In [79]:
chars_s.groupby(['gender', 'tier']).size()

Series([], dtype: int64)

In [80]:
chars_s.groupby(['tier', 'gender']).size()

Series([], dtype: int64)

이런.. 특수 npc는 인기도도 없습니다.

인기는 많을텐데...


### 인기도의 종별 분포?

종은 외모에 상당히 결정적입니다.


--> 뭔가 경향이 있긴 한데 종이 너무 많아서 도저히 알 수가 없다.


In [84]:
chars_v.groupby(['species', 'tier']).size()
print(_.to_string())

species  tier
개        2.0      2
         3.0      3
         4.0      4
         5.0      4
         6.0      3
개구리      2.0      1
         3.0      1
         4.0      1
         5.0      8
         6.0      7
개미핥기     5.0      3
         6.0      4
고릴라      5.0      1
         6.0      8
고양이      1.0      2
         2.0      6
         3.0      4
         4.0      4
         5.0      5
         6.0      2
곰        4.0      3
         5.0      3
         6.0      9
꼬마곰      1.0      3
         2.0      1
         3.0      1
         4.0      2
         5.0      5
         6.0      4
늑대       2.0      2
         3.0      4
         4.0      3
         5.0      2
다람쥐      1.0      1
         2.0      1
         3.0      2
         4.0      4
         5.0      6
         6.0      4
닭        5.0      4
         6.0      5
독수리      2.0      1
         5.0      3
         6.0      5
돼지       4.0      1
         5.0      1
         6.0     12
말        2.0      1
         3.0      1
      

In [87]:
chars_v.groupby(['tier', 'species']).size()
print(_.to_string())

tier  species
1.0   고양이         2
      꼬마곰         3
      다람쥐         1
      문어          2
      사슴          1
      염소          1
      코뿔소         1
      토끼          1
2.0   개           2
      개구리         1
      고양이         6
      꼬마곰         1
      늑대          2
      다람쥐         1
      독수리         1
      말           1
      사슴          3
      양           2
      오리          2
      타조          1
      펭귄          1
3.0   개           3
      개구리         1
      고양이         4
      꼬마곰         1
      늑대          4
      다람쥐         2
      말           1
      문어          1
      사슴          3
      염소          2
      코끼리         1
      코알라         1
      토끼          2
      햄스터         3
4.0   개           4
      개구리         1
      고양이         4
      곰           3
      꼬마곰         2
      늑대          3
      다람쥐         4
      돼지          1
      말           1
      사슴          3
      사자          1
      악어          2
      양           1
      염소          2
      